### PR TITLE
Consistency tests

### PR DESCRIPTION
**Pull request by lintusj1**
Adds tests for consistent results given the same seed and initial conditions. Includes quite a bit of refactoring.

### CODE REVIEW

File: elfi/methods/parameter_inference.py [hunk 1]
```diff
@@ -733,152 +714,207 @@ def current_population_threshold(self):
...
+                                        exploration_rate=exploration_rate,
+                                        seed=self.seed)
+
+        self.n_initial_evidence = n_initial
+        self.n_precomputed_evidence = n_precomputed
+        self.update_interval = update_interval
 
-        # Fix bounds of user-supplied target_model
-        if type(target_model.bounds) == dict:
-            target_model.bounds = [target_model.bounds[k] for k in model.parameter_names]
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: Why remove this?

File: elfi/methods/parameter_inference.py [hunk 2]
```diff
@@ -733,152 +714,207 @@ def current_population_threshold(self):
...
+        else:
+            precomputed = initial_evidence
+            n_initial_evidence = len(precomputed[self.target_name])
+
+        if n_initial_evidence < 0:
+            raise ValueError('Number of initial evidence must be positive or zero '
+                             '(was {})'.format(initial_evidence))
+        elif n_initial_evidence < n_required:
+            logger.warning('We recommend having at least {} initialization points for '
+                           'reliable initialization (now {})'\
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: Maybe drop 'reliable'? :)

File: elfi/methods/parameter_inference.py [hunk 3]
```diff
@@ -733,152 +714,207 @@ def current_population_threshold(self):
...
-        self.acquisition_method = acquisition_method or \
-                                  LCBSC(target_model, prior=ModelPrior(self.model),
-                                        noise_cov=acq_noise_cov, seed=self.seed)
-        # TODO: move some of these to objective
-        self.n_evidence = initial_evidence
-        self.target_model = target_model
-        self.n_initial_evidence = initial_evidence
-        self.update_interval = update_interval
+            n_initial_evidence = n_required
+        elif isinstance(initial_evidence, (int, np.int)):
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: integer float?

File: elfi/methods/bo/acquisition.py [hunk 4]
```diff
@@ -75,61 +73,70 @@ def evaluate_gradient(self, x, t=None):
...
+        x = self._add_noise(x)
 
         return x
 
+    def _add_noise(self, x):
+        # Add noise for more efficient fitting of GP
+        if self.noise_var is not None:
+            noise_var = np.asanyarray(self.noise_var)
+            if noise_var.ndim == 0:
+                noise_var = np.tile(noise_var, self.model.input_dim)
```
Comment (lintusj1) [inline elfi/methods/bo/acquisition.py]: I guess both would work. Doing it dynamically you don't have to worry about the value being changed -> simpler code.

File: elfi/methods/parameter_inference.py (hunk 1 above)
Comment (lintusj1) [inline elfi/methods/parameter_inference.py]: Why would not the bounds be correct in the user-supplied model? They are required in the constructor.

Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: I think Antti already had a problem with this once

File: elfi/methods/bo/acquisition.py (hunk 4 above)
Comment (vuolleko) [inline elfi/methods/bo/acquisition.py]: Not good to repeatedly apply the same operations on constants. I suppose there should be a setter to allow dynamic changes, if such are needed.

File: elfi/methods/parameter_inference.py (hunk 1 above)
Comment (lintusj1) [inline elfi/methods/parameter_inference.py]: I think it's clearer that no changes are done to the user supplied model.

**Pull request merged**